feat(fulfillment): support postal_expression filtering for geo zone matching

Saryazdi-Saman opened the pull request:
🚀 Adds postal code-based zone matching for improved fulfillment filtering.

## Summary

This PR introduces support for filtering geo zones based on the `postal_expression` field when listing shipping options for a given address. It enables both exact and prefix-based postal code matching to improve fulfillment granularity.

Previously, geo zone matching only supported country, province, and city fields. There was no way to match based on full or partial postal codes.
This made it difficult to define fine-grained service areas or neighbourhoods using real-world postal code patterns.

With this change:
- Exact postal code matches (`matches` field) and postal prefix matches (`starts_with` field) are now supported.
- Shipping options are matched only if the service zone fully contains the provided address. (e.g., a shippingOption serving Toronto/Canada won't be matched with an address in Montreal/Canada just because they share the same Country)
- Improves flexible, efficient postal code-based fulfillment territory management.

## Changes Introduced

- Added `PostalExpression` and `PostalExpressionFilterableProps` interfaces for structured postal filtering.
- Extended `buildGeoZoneConstraintsFromAddress` to generate dynamic `$or` conditions for postal expressions.
- Injected the `type` field (e.g., `zip`, `city`) into constraints to improve matching precision.
- Added GIN index on `geo_zone.postal_expression` (JSONB) for optimized lookup.
- Added integration tests validating:
  - Exact postal code matching
  - Prefix postal code matching
  - Non-matching behaviour

## Testing

- Added integration tests for exact and prefix postal code scenarios
- Verified service zone matching works correctly for both direct and partial postal codes
- Confirmed that shipping options which don't fully cover an area are excluded from results (This scenario would have failed prior to this PR)
- All tests pass locally and reflect expected behaviour

## Notes

- Assumes PostgreSQL is being used, as it leverages JSONB operators and GIN indexing

Saryazdi-Saman committed:
chore(types): add JSONB filter interfaces and postal expression structure

- Introduce `JsonbObject` as a base flexible JSONB type
- Add `FilterableJSONB<T>` and `FilterableJSONBArray<T>` for advanced query filtering
- Define `PostalExpression` and `PostalExpressionFilterableProps` for geo zone filtering
- Update `FilterableGeoZoneProps` to support postal_expression filtering

Saryazdi-Saman committed:
feat(fulfillment): enable postal_expression filtering in geo zone matching

- Update `buildGeoZoneConstraintsFromAddress` to support filtering by postal code
- Inject `type` field  constraint for geo zone  matching
- Generate JSONB filters for the `postal_expression` field
- Create `$or` conditioned for exact and prefix-based postal code comparisons

Saryazdi-Saman committed:
test(fulfillment): add tests for postal_expression based geo zone matching

- Validate listing shipping options with exact postal_expression matches
- Validate listing shipping options using postal_expression prefix matching
- Confirm no matches returned for mismatched postal codes
- Confirm behavior for non-postal_expression based geo zones (e.g., city matching)

Saryazdi-Saman committed:
chore(migration): add GIN index for postal_expression in fulfillment module

Saryazdi-Saman commented:
Thanks for reviewing! Happy to adjust based on feedback.

adrien2p reviewed (approved):
Excellent, that looks good to me but I ll wait for @olivermrbl opinion as well

olivermrbl commented on packages/core/types/src/fulfillment/common/geo-zone.ts:
```diff
@@ -90,5 +121,49 @@ export interface FilterableGeoZoneProps
+// /**
+//  * Filter structure for JSONB arrays.
+//  *
+//  * Supports containment queries to check if the array contains specific elements
+//  * or is fully contained within another array.
+//  */
+// export interface FilterableJSONBArray<T> {
+//   /**
+//    * Matches JSONB arrays that contain all of the specified elements.
+//    */
+//   $contains?: T[]
+
+//   /**
+//    * Matches JSONB arrays that are fully contained within the specified elements.
+//    */
+//   $contained?: T[]
+// }
+
+// /**
+//  * Generic JSONB object structure.
+//  * Used to model flexible JSONB fields in the database, allowing flexible filtering strategies.
+//  */
+// export interface JsonbObject {
+//   [key: string]: unknown
+// }
+
+// /**
+//  * Filter structure for a JSONB object, supporting key existence queries.
+//  */
+// export interface FilterableJSONB<T extends JsonbObject> {
+//   /**
+//    * Matches JSONB objects that contain all the specified keys.
+//    */
+//   $hasKeys?: (keyof T)[]
+
+//   /**
+//    * Matches JSONB objects that contain at least one of the specified keys.
+//    */
+//   $hasSomeKeys?: (keyof T)[]
+// }
```
question: can we remove this, or is there a reason to keep it here?

Saryazdi-Saman replied:
Good catch — this was leftover from a previous iteration. Cleaning it up now 🙌

olivermrbl reviewed (approved):
LGTM, appreciate the contribution.

olivermrbl committed:
Create pretty-trees-march.md

adrien2p commented on packages/modules/fulfillment/src/services/fulfillment-module-service.ts:
```diff
@@ -2419,17 +2419,71 @@ export default class FulfillmentModuleService
+   * Builds a postal expression filter with exact and prefix matching conditions
+   * based on the provided postal code.
+   */
+  protected static buildPostalExpressionFilter(postalCode: string) {
+    return [
+      { postal_expression: { matches: { $contains: [postalCode] } } },
+      ...Array.from(
+        { length: Math.max(0, Math.min(postalCode.length - 1, 4)) },
+        (_, i) => ({
+          postal_expression: { starts_with: { $contains: [postalCode.slice(0, i + 2)] } }
+        })
+      ),
+    ]
+  }
```
```suggestion
protected static buildPostalExpressionFilters(postalCode: string) {
  const postalExpressions = [];

  postalExpressions.push({
    postal_expression: {
      matches: { $contains: [postalCode] }
    }
  });

  // Add partial match expressions for first 2-5 characters
  // We start with at least 2 characters and add more if available
  const minPartialLength = 2;
  const maxPartialLength = Math.max(0, Math.min(postalCode.length - 1, 5));
  
  for (let length = minPartialLength; length <= maxPartialLength; length++) {
    const partialCode = postalCode.slice(0, length);
    postalExpressions.push({
      postal_expression: {
        starts_with: { $contains: [partialCode] }
      }
    });
  }

  return postalExpressions;
}
```

Suggestion to make it more human friendly and simplify some computation that can be controlled differently and make it easier to evolve for potential future updates 👍 let me know what you think

Saryazdi-Saman replied:
Great suggestion — I agree this is more readable and easier to maintain. I’ve pushed the update 🙌

Saryazdi-Saman committed:
refactor(postal_expression): simplify filter builder for better readability

- Replaced Array.from with clearer for-loop structure
- Improves maintainability and clarity of logic

Co-authored-by: Adrien de Peretti <adrien.deperetti@gmail.com>

Saryazdi-Saman committed:
chore(cleanup): remove commented-out JSONB filter interfaces

olivermrbl committed:
Merge branch 'develop' into feat/postal-expression-matching

Saryazdi-Saman committed:
fix(fulfillment-service): correct postal expression filter reference and type assignment

Saryazdi-Saman committed:
Merge branch 'develop' into feat/postal-expression-matching

Saryazdi-Saman commented:
I noticed that the last commit had caused a TypeScript build error (due to a function name mismatch and a missing type on the postalExpressions array).

I’ve fixed both issues in the latest commit — build now passes locally and in CI ✅

Saryazdi-Saman commented:
Hey @adrien2p @olivermrbl — just looping you in since I pushed a quick fix last night to resolve a build issue. CI is all green now, let me know if it needs another approval. Thanks!

adrien2p committed:
Merge branch 'develop' into feat/postal-expression-matching

Saryazdi-Saman commented:
Hey team! Just checking in  — I saw this was approved (thank you!), and I wanted to see if there’s anything else you’d like me to update or if it’s just waiting for the right time to merge.

I’m also about to start working on the related admin UI for service zone management. 
Would you prefer that I add the UI changes to this same PR, or create a separate one building on top of this branch? Happy to do whichever makes more sense for your workflow.

Saryazdi-Saman committed:
Merge branch 'develop' into feat/postal-expression-matching
